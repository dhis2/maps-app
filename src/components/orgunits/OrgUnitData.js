import { useDataEngine } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { CircularLoader } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import PeriodSelect from '../periods/PeriodSelect.js'
import styles from './styles/OrgUnitData.module.css'

/*
 *  Displays a period selector and org unit data items (data elements, indicators, reporting rates, program indicators)
 */
const OrgUnitData = ({ id, periodType, defaultPeriod, defaultItems }) => {
    const [period, setPeriod] = useState(defaultPeriod)
    const [items, setItems] = useState(defaultItems)
    const [isLoading, setIsLoading] = useState(false)
    const engine = useDataEngine()

    // Load data items if period is changed
    useEffect(() => {
        const query = {
            orgUnitProfile: {
                resource: `organisationUnitProfile/${id}/data`,
                params: {
                    period: period.id,
                },
            },
        }

        const fetchOrgUnitProfile = async () => {
            setIsLoading(true)
            const { orgUnitProfile } = await engine.query(query)
            setItems(orgUnitProfile.dataItems)
            setIsLoading(false)
        }
        if (period.id === defaultPeriod.id) {
            setItems(defaultItems)
        } else {
            fetchOrgUnitProfile()
        }
    }, [id, period, defaultPeriod, defaultItems, engine])

    return (
        <div className={styles.orgUnitData}>
            <PeriodSelect
                label={null}
                periodType={periodType}
                period={period}
                onChange={setPeriod}
                className={styles.periodSelect}
            />
            <div className={styles.dataTable}>
                {isLoading && (
                    <div className={styles.loadingMask}>
                        <CircularLoader />
                    </div>
                )}
                {Array.isArray(items) && items.length ? (
                    <table>
                        <tbody>
                            {items.map(({ id, label, value }) => (
                                <tr key={id}>
                                    <th>{label}</th>
                                    <td>{value}</td>
                                </tr>
                            ))}
                        </tbody>
                    </table>
                ) : (
                    <div className={styles.noData}>
                        {i18n.t('No data found for this period.')}
                    </div>
                )}
            </div>
        </div>
    )
}

OrgUnitData.propTypes = {
    defaultPeriod: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    periodType: PropTypes.string.isRequired,
    defaultItems: PropTypes.array,
}

export default OrgUnitData
