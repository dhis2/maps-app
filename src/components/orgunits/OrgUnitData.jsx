import { useDataQuery } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { CircularLoader } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import {
    getFixedPeriodsByType,
    filterFuturePeriods,
} from '../../util/periods.js'
import PeriodSelect from '../periods/PeriodSelect.jsx'
import styles from './styles/OrgUnitData.module.css'

const ORGUNIT_PROFILE_QUERY = {
    profile: {
        resource: 'organisationUnitProfile',
        id: ({ id }) => `${id}/data`,
        params: ({ period }) => ({
            period,
        }),
    },
}

// Only YEARLY period type is supported in first version
const periodType = 'YEARLY'
const year = String(new Date().getFullYear())
const periods = getFixedPeriodsByType({ periodType, year })
const defaultPeriod = filterFuturePeriods(periods)[0] || periods[0]

/*
 *  Displays a period selector and org unit data items
 * (data elements, indicators, reporting rates, program indicators)
 */
const OrgUnitData = ({ id }) => {
    const [period, setPeriod] = useState(defaultPeriod)
    const { loading, data, refetch } = useDataQuery(ORGUNIT_PROFILE_QUERY, {
        lazy: true,
    })

    useEffect(() => {
        if (id && period) {
            refetch({
                id,
                period: period.id,
            })
        }
    }, [id, period, refetch])

    return (
        <div className={styles.orgUnitData} data-test="org-unit-data">
            <PeriodSelect
                label={null}
                periodType={periodType}
                period={period}
                onChange={setPeriod}
                className={styles.periodSelect}
            />
            <div className={styles.dataTable}>
                {loading && (
                    <div className={styles.loadingMask}>
                        <CircularLoader />
                    </div>
                )}
                {Array.isArray(data?.profile.dataItems) &&
                data.profile.dataItems.length ? (
                    <table data-test="org-unit-data-table">
                        <tbody>
                            {data.profile.dataItems.map(
                                ({ id, label, value }) => (
                                    <tr key={id}>
                                        <th>{label}</th>
                                        <td>{value}</td>
                                    </tr>
                                )
                            )}
                        </tbody>
                    </table>
                ) : (
                    <div className={styles.noData}>
                        {i18n.t('No data found for this period.')}
                    </div>
                )}
            </div>
        </div>
    )
}

OrgUnitData.propTypes = {
    id: PropTypes.string.isRequired,
}

export default OrgUnitData
