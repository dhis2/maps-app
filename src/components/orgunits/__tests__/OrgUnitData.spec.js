import { render } from '@testing-library/react'
import React from 'react'
import OrgUnitData from '../OrgUnitData.js'

/* eslint-disable react/prop-types */
jest.mock('@dhis2/ui', () => {
    const originalModule = jest.requireActual('@dhis2/ui')

    return {
        __esModule: true,
        ...originalModule,
        SingleSelectField: function Mock(props) {
            return <div className="ui-SingleSelectField">{props.children}</div>
        },
        SingleSelectOption: function Mock(props) {
            return <div className="ui-SingleSelectOption">{props.children}</div>
        },
        Tooltip: function Mock(props) {
            return <div className="ui-Tooltip">{props.children}</div>
        },
        Button: function Mock(props) {
            return <div className="ui-Button">{props.children}</div>
        },
    }
})
/* eslint-enable react/prop-types */

const defaultProps = {
    id: 'DiszpKrYNg8',
    periodType: 'YEARLY',
    defaultPeriod: {
        id: '2021',
    },
}

const defaultItems = [
    {
        id: 'WUg3MYWQ7pt',
        label: 'Total Population',
        value: 3503,
    },
    {
        id: 'DTVRnCGamkV',
        label: 'Total population < 1 year',
        value: 140,
    },
    {
        id: 'vg6pdjObxsm',
        label: 'Population of women of child bearing age (WRA)',
        value: 716,
    },
    {
        id: 'Uvn6LCg7dVU',
        label: 'ANC 1 Coverage',
        value: 368.2,
    },
    {
        id: 'eTDtyyaSA7f',
        label: 'FIC <1y',
        value: 291.4,
    },
]

describe('Org unit data items', () => {
    it('renders with the default period and items', () => {
        const { container } = render(
            <OrgUnitData {...defaultProps} defaultItems={defaultItems} />
        )
        expect(container).toMatchSnapshot()
    })
})
