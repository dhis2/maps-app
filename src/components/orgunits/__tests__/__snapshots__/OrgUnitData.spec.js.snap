// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Org unit data items renders with the default period and items 1`] = `
<div>
  <div
    class="orgUnitData"
  >
    <div
      class="periodSelect periodSelect"
    >
      <div
        class="inputField select"
      >
        <div
          class="ui-SingleSelectField"
        >
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
          <div
            class="ui-SingleSelectOption"
          />
        </div>
      </div>
      <div
        class="stepper"
      >
        <div
          class="ui-Tooltip"
        >
          <div
            class="ui-Button"
          />
        </div>
        <div
          class="ui-Tooltip"
        >
          <div
            class="ui-Button"
          />
        </div>
      </div>
    </div>
    <div
      class="dataTable"
    >
      <table>
        <tbody>
          <tr>
            <th>
              Total Population
            </th>
            <td>
              3503
            </td>
          </tr>
          <tr>
            <th>
              Total population &lt; 1 year
            </th>
            <td>
              140
            </td>
          </tr>
          <tr>
            <th>
              Population of women of child bearing age (WRA)
            </th>
            <td>
              716
            </td>
          </tr>
          <tr>
            <th>
              ANC 1 Coverage
            </th>
            <td>
              368.2
            </td>
          </tr>
          <tr>
            <th>
              FIC &lt;1y
            </th>
            <td>
              291.4
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
